### ISSUE TITLE
Discussion on general level of PEP8 compliance. 

### ISSUE THREAD

**Issue by alexggmatthews** [GPflow/GPflow#459]
Branched from #452 which is a more specific question at the suggestion @javdrher 

I've personally always felt that extreme adherence to a coding standard could slow down development. I'm pretty happy with the compromise level we've existed at since inception which really has come from a common sense agreement between the main developers. 

I recognise as the GPflow community grows and branches we need to keep these things under review.  

Despite the fact the GPflow community is growing I still think there is a danger of distracting effort and attention from more importance questions about code content. 

I don't think we should block PRs unless they have drastically the wrong coding style. There are after all automated tools we could periodically run if people wish.

The specific case of variable names mentioned by @javdrher in #452 is a good example. I'm totally happy for variable names to be chosen to match papers if that it was the contributors feel is most sensible.

To summarize, I think coding standards such as PEP8 can be helpful in a project of our size, but extreme adherence for its own sake it literally style over substance. 

Thoughts?

**Comment by awav** [comment]
Hello everyone!

The code style guides exist to simplify maintenance, get better readability, hence make your life easier as contributor and whole community as well. Different tools can relief development process, something like `autopep8`, `pep8fy`, `YAPF`, plugins to editors: `python-pylint`, `python-tools` and etc.
@javdrher gave extremely good example, but various authors can have divergent preferences for variable names, e.g. `Xnew`, `xNew`,  `x_New`, such flexibility makes code look like a mess. All of these name versions can be adapted as either `xnew` or `x_new` without loss of meaning. Most important part here is a context - why this variable was called in this way, if it is a replication of paper codestyle, where is a paper link and short description what it means. Here is an example from kernels.py:
```
    @AutoFlow((float_type, [None, None]))
    def compute_K_symm(self, X):
        return self.K(X)

    @AutoFlow((float_type, [None, None]), (float_type, [None, None]), (float_type,))
    def compute_eKxz(self, Z, Xmu, Xcov):
        return self.eKxz(Z, Xmu, Xcov)

    @AutoFlow((float_type, [None, None]), (float_type, [None, None]), (float_type, [None, None, None, None]))
    def compute_exKxz(self, Z, Xmu, Xcov):
        return self.exKxz(Z, Xmu, Xcov)
```
First of all, I think names are good, sometimes it does makes sense to ignore rules and write concise code (using abbreviations or whatever). But here it lacks a description of what's happening within `compute_exKxz`, `compute_eKxz`, `compute_exKxz`, and why autoflow has such parameters.

In any case, `pylint` has python directives, in fact it is just comments, with special labels for neglecting warning/error/info messages and it also has configuration file, where you can relax some rules and say that camelcase style is acceptable by GPflow. For example increasing line length from 80 to 100 is totally okay for me. If someone is going to use non-standard names, he/she could use directives to silence messages.

**Comment by awav** [comment]
I think, it is important topic to discuss, so others are welcome to share their opinion.
@jameshensman, @icouckuy, @javdrher, @Bonnevie, and many more :)

**Comment by markvdw** [comment]
I'm very much for the increasing the line length from 80 to 100. We don't all work on CGA terminals any more.

**Comment by alexggmatthews** [comment]
@awav @markvdw  can discuss this over lunch on Monday as well. Some more thoughts:

I think the decision to use an automated code tool as part of the build process would need to be taken very carefully. Of course, no automated code tool is perfect....

From a machine learning perspective any automated code tool is a probabilistic decision process, governed by probabilistic decision theory. To fully specify the problem we will need a loss and the loss that different stake holders in the code would use is not necessarily the same. For instance some people will care deeply that we have `Xnew, xNew, x_New,` in the code and some people will not care at all. Any decision that is made by the community will need to attempt to consider the range of views on this. We could also think about it in terms of developer time. Is the amount of developer time saved by using a given tool higher or lower? This will depend on the quality of the tool (how many false positives and negatives for a given user's definition of useful) and how well it is used.

For a more practical perspective, I've seen much vaunted automated code tools badly implemented before to the extent of drastically slowing down development. This has taught me to exercise a healthy scepticism on such things in general and where used insist they are implemented well,  _before they become an established part of the workflow_. 

Final thought: I think it is important that whatever conclusion we come to we don't alienate potential contributors from outside the core group by adding unnecessarily to the barriers to entry. It is not a saving in developer time if we put off a large proportion of our potential developers. GPflow has always tried to be a very open community and the project has greatly benefited from this. 

I'd encourage everyone who has an opinion on this to express it :)

**Comment by icouckuy** [comment]
I love your Bayesian analysis of the problem :) I'm not familiar with any automated code tools for PEP8. I use pycharm which has continuous checking of PEP8 violations which I try to adhere to as much  as possible.

In the end it depends on the goals of GPflow. First I thought it was best to comply with the paper notation as Matthews mentioned which makes it easier for newcomers to cross-check an approach with the corresponding paper and try to understand it. In short, keep only the code tying everything together in PEP8 and the algorithms keep the notation of the paper, most notably the predictive distributions and likelihood lowerbounds.

However, as far as I know the goal of GPflow was not to be a learning tool (in contrast to GPy). In addition, papers of extension to the same methodologies might even have significant notational differences. This leads me more towards trying to adhere to the PEP8 standard.

I do not care that much for it that I will fight over it though. In my opinion it is crucial to keep the contributing guidelines (contributing.md) as short as possible so newcomers can easily create new PRs. Code cleanup can then happen in the PR using common sense and/or a pep8 tool, as is done now.

**Comment by javdrher** [comment]
I believe the analysis of @alexggmatthews is going to be confirmed empirically by the diversity in responses in this discussion. My opinion divert from Ivo's. I'm also using pycharms PEP8 and I fully agree most of the directives are plain common sense in terms of code readability and maintainability. Actually, my first programming steps were in PHP4: after one week of browsing examples online I understood the message how important some fundamental style requirements like spaces after commas and surrounding operators are. No discussion there.

For the naming aspect, I consider PEP8 useful recommendations but to be ignored when appropriate. In terms of package naming, classes and methods I can still follow to some level, but I think the constraints on variable naming are a very bad idea for two reasons:
- even without caps, there are plenty of PEP8 legal ways to make naming a complete mess which means the need for common sense isn't ruled out.  
- sensible variable naming affects code readability a lot. Developers mastering the art of condensing the semantics of variables in a short name are exceptional (I'm horrible at it) but reading and maintaining their code is a lot easier. Therefor if a name is better of as a violation it should be. As a consequence I think reflecting common symbols in papers (this is of course not universal but X/Y/F/Z/q(.)/K ring bells) isn't a matter of being a learning tool. It is about code quality and readability.

A bit more practical: I know we can configure the policies of linters, or place pylint directives in the code to suppress linter warnings, but on the other hand I'm convinced contributors are not massively violating PEP8 up to the point it becomes confusing/a problem. Therefor I'm in doubt integrating automated quality checks will actually help the project (in its current state) forward but I'm open to be proven otherwise. I fully agree with Matthews though that there is a real threat of being worse of with slower development and a harder level to contribute. Occasionally (e.g., before a release), running a linter and resolving style problems (e.g. whitespaces) which have been overlooked like @awav did could be a middle ground?

**Comment by alexggmatthews** [comment]
Thank you for your contributions everyone. I sat down with @awav @marvdw and @jameshensman over lunch today to make a plan for this. We took into account the helpful contributions on this forum from those who aren't in the UK so thank you for taking the time to contribute.

The plan is:
1) To create, in consultation with the community, a style file for `pylint`. 
2) To run `pylint` with this style file on pull requests. To start with this will only be as `advice' for reviewers and will not be a requirement to merge. Reviewers will still use their discretion on style which is what happens at the moment.
3) Periodically we will from time to time do PRs which solely cover style. To start with these will be reviewed like any other PR though it is likely we will relax this.
4) If after a trial period of time we feel that the continuous integration advice has been useful and reliable we will, in consultation with the GPflow community, make it a requirement of merging. This would save the need to do 3) . If it turns out that `pylint` can never be made reliable enough to be fully automatic we won't make it a requirement and will keep it as an advisory diagnostic.
